### PR TITLE
Add support for movie file names containing AKA

### PR DESCRIPTION
**Pull request by 0aeonnoea0**
#### Database Migration
NO

#### Description
First pass at addressing #3748. Have hooked everything up and it seems to be working well when testing on my server.

This PR adds an `AlternativeMovieTitles` to `ParsedMovieInfo` which contains alternative titles specified by the occurrence of `AKA`. An example of this is `Akahige.AKA.Red.Beard.1965.CD1.CRiTERiON.DVDRip.XviD-KG.avi`, which parses to alternative titles `Akahige` and `Red Beard`, and movie title `Akahige AKA Red Beard`.

I've added a bunch of test cases from files I've found on PTP, hopefully there's a good enough variety there.

My biggest question is whether it's the best idea to turn `MovieTitle` into a collection rather than adding an additional property? I'm worried that by having an additional property to check we'll end up adding complexity that could be avoiding by just making everything search the whole collection of names.

#### Todos
- [X] Parser support and relevant test fixtures
- [x] Hooking this up to the wider codebase
- [ ] Ensure the frontend works correctly and add further tests

#### Issues Fixed or Closed by this PR
* #3748

### CODE REVIEW

File: src/NzbDrone.Core.Test/ParserTests/ParserFixture.cs [hunk 1]
```diff
@@ -60,6 +60,56 @@ public void should_parse_movie_title(string postTitle, string title)
             Parser.Parser.ParseMovieTitle(postTitle, true).MovieTitle.Should().Be(title);
         }
 
+        [TestCase(
+            "L'hypothèse.du.tableau.volé.AKA.The.Hypothesis.of.the.Stolen.Painting.1978.1080p.CINET.WEB-DL.AAC2.0.x264-Cinefeel.mkv",
+            "L'hypothèse du tableau volé AKA The Hypothesis of the Stolen Painting",
+            new string[] { "L'hypothèse du tableau volé", "The Hypothesis of the Stolen Painting" })]
```
Comment (0aeonnoea0) [inline src/NzbDrone.Core.Test/ParserTests/ParserFixture.cs:66]: Is there a slightly neater way to set up these test cases? I haven't used C# in ages so I'm a bit out of the loop!

File: src/NzbDrone.Core/MediaFiles/MovieImport/Manual/ManualImportService.cs [hunk 2]
```diff
@@ -143,7 +143,7 @@ private ManualImportItem ProcessFile(string rootFolder, string baseFolder, strin
 
             if (movie == null)
             {
-                _parsingService.GetMovie(relativeFile.Split('\\', '/')[0]);
+                movie = _parsingService.GetMovie(relativeFile.Split('\\', '/')[0]);
```
Comment (0aeonnoea0) [inline src/NzbDrone.Core/MediaFiles/MovieImport/Manual/ManualImportService.cs:146]: This was previously not doing anything

File: src/NzbDrone.Core/NetImport/RSSImport/RSSImportParser.cs [hunk 3]
```diff
@@ -146,7 +146,7 @@ protected virtual Movie ProcessItem(XElement item, Movie releaseInfo)
 
             if (result != null)
             {
-                releaseInfo.Title = result.MovieTitle;
+                releaseInfo.Title = result.PrimaryMovieTitle;
```
Comment (0aeonnoea0) [inline src/NzbDrone.Core/NetImport/RSSImport/RSSImportParser.cs:149]: Wasn't sure what else would make sense here with multiple possible titles - should I fill in `AlternativeTitles`?

File: src/NzbDrone.Core/Parser/ParsingService.cs [hunk 4]
```diff
@@ -121,14 +122,56 @@ public Movie GetMovie(string title)
...
+            return null;
+        }
+
+        private Movie LookupParsedMovie(ParsedMovieInfo parsedMovieInfo)
+        {
+            var movie = LookupParsedMovie(parsedMovieInfo, requireExactYear: true);
+
+            if (movie == null)
             {
-                movies = _movieService.FindByTitle(parsedMovieInfo.MovieTitle.Replace("DC", "").Trim());
```
Comment (0aeonnoea0) [inline src/NzbDrone.Core/Parser/ParsingService.cs:128]: I looked into this random removal of "DC" and it comes from 52fb29ee189e7da807729683840097c58ef8ebc2 with extremely little explanation. My guess is this to do with edition information - should I move this into the Parser itself? I'm not sure how best to keep it in here.

**Comment by sevospl** [comment]
Hi! First of all, thanks for this PR, it'd be very useful for my use case. I'm not quite sure if it's already been merged into v3 already but I'd like to bring it up again if I may.

There's one tracker that names torrents like that, here's an example:

`Nagi instynkt / Basic.Instinct.1992.DIRECTOR'S.CUT.MULTi.720p.BluRay.x264.DTS.AC3-BiRD`

So instead of `AKA` they're using a slash to differentiate 2 language names. Which probably would cause issues.

I would appreciate taking it into consideration. Thanks! :)

**Comment by fryfrog** [comment]
The only movie I could think of w/ a `/` in the name was `Face/Off` and it looks like it has *almost* no releases w/ the `/` in it, but I did find *one*. So if you do add `/` as a separator like `AKA`, it'd probably be a good test case.

```
Face/Off 1997 Multisub 1080p Blu-ray Remux VC-1 LPCM 5.1 - KRaLiMaRKo
```

**Comment by 0aeonnoea0** [comment]
Hey @sevospl, glad this PR might be useful for you! At the moment it's not merged - it's pretty much ready to go, I just need to finish off by writing a few more tests. Have had an unexpectedly hectic couple of weeks - hopefully I can sort this in the coming weekend.

I think that the `/` format should be able to be made to work - the way that this change works is that the parser has multiple possible titles for a movie rather than just one when it looks at a filename. I then go through this list of titles sequentially when trying to match a movie. The multiple titles allows us to avoid issues where we have `AKA` (or `/`) in the movie name, because (for example) `Foreign.Name.AKA.Local.Name` becomes `["Foreign Name AKA Local Name", "Foreign Name", "Local Name"]`. That way we avoid just randomly dropping a possible title - and would avoid issues in the case @fryfrog highlighted.

I wonder if you might be able to supply me with some further examples of releases with `/` in them? I'd be keen to add some test cases so I can get the behaviour right.

One possible concern I could see is a movie with `/` or `AKA` in the name, plus a release name with multiple titles (e.g. `Face/Off AKA Foreign/Name`) - which do I prioritise splitting on? is this even a likely occurrence worth covering?

**Comment by fryfrog** [comment]
I think splitting *only* by `/` *or* `AKA` is a good start and probably all that'd ever be needed. I'd *probably* prioritize `AKA`, only using `/` if `AKA` wasn't present.

**Comment by sevospl** [comment]
@0aeonnoea0 A few more examples:
 Pod ciemnymi gwiazdami / Beast.2017.MULTi.1080p.BluRay.x264.DTS.AC3
Jaś Fasola: Nadciąga totalny kataklizm / Bean.1997.MULTi.720p.BluRay.x264
Battleship: Bitwa o Ziemię / Battleship.2012.MULTi.720p.BluRay.x264.DTS.AC3
Nagi instynkt / Basic.Instinct.1992.DIRECTOR'S.CUT.MULTi.720p.BluRay.x264.DTS.AC3
13 Dzielnica / Banlieue.13.2004.MULTi.1080p.BluRay.x264.DTS.AC3
Ostatnie zlecenie / Bangkok.Dangerous.2008.MULTi.720p.BluRay.x264.DTS.AC3

**Comment by sevospl** [comment]
@0aeonnoea0 Could you please extend the PR to support my specific case scenario (slash)?

**Comment by Qstick** [comment]
@JBou want to look this over?

**Comment by JBou** [comment]
I currently don't have much time, have to finish a work the next weeks. But maybe i find some time, let's talk on discord.

What about splitting not only by "/" or "AKA", but instead " / " or "./." or " AKA " or ".AKA."
Using regex this would be possible with word boundaries (\b). This way, Face/Off would not be splitted, because there are no surrounding spaces.

**Comment by Qstick** [comment]
@0aeonnoea0 care to rebase this?

**Comment by JohnCytron** [comment]
Love to see this implementation.

Some other examples with `A.K.A.`:
-  `Kynodontas a.k.a. Dogtooth 2009 1080p Blu-ray Remux AVC DTS-HD MA 5.1 - KRaLiMaRKo`
- `Die Welle a.​k.​a.​ The Wave 2008 1080p FRA Blu-​ray AVC DTS-​HD MA 5.​1-​F13@HDSpace`
- `Kimi to Nami ni Noretara a.​k.​a.​ Ride Your Wave 2019 Repack 1080p Blu-​ray Remux AVC DTS-​HD MA 5.​1-​E.​N.​D`

**Comment by Qstick** [comment]
Yea I'm going to pick this up after the lists UI rework is merged, should be in Aphrodite in short order

**Comment by Qstick** [comment]
Replaced by #5221

**Pull request closed without merging**